### ISSUE TITLE
Loss function

### ISSUE THREAD

**Issue by mikebawol** [experiencor/basic-yolo-keras#39]
I am attempting to implement a custom version of this algorithm and I can't wrap my head around the loss function, I've read the paper many times and it seems as though everyone's interpretation of this function differs. Can someone explain to me in simple terms what is occurring in this function? I see sigmoid functions and exponentials, while neither of those appear in the original paper.

**Comment by experiencor** [comment]
1. You will find sigmoid and exponential in the YOLOv2 (YOLO9000: Better, Faster, Stronger), not YOLOv1 (You Only Look Once: Unified, Real-Time Object Detection). FYI, I am implementing YOLOv2.

2. The basic idea is that the loss function can be broken into 4 parts: position loss xy, size loss wh, confidence loss and classification loss. These losses simply penalize the wrong predictions of position, size, confidence and classification. All losses are mean squared errors, except classification loss, which uses cross entropy function. 

3. The parts of the code which compute the masks and warm training are not included in the paper. They are found in the C implementation of YOLO (https://github.com/pjreddie/darknet). The code still works without these tricks, but the tricks do boost up performance.

Hope it help.

**Comment by bshao001** [comment]
@experiencor I am also having trouble understanding the loss function. Could you please briefly explain why masks are applied in the loss function? Thank you so much!

**Comment by experiencor** [comment]
The mask determines what bounding boxes to be penalized. 1 for penalizing and 0 for not. For example, we only penalize the coordinate predictions of the bounding boxes which are responsible for predicting certain objects. So the coordinate mask has the value of 1 at the predictor bounding boxes and 0 at others. In the final loss, the mask is multiplied to the difference between the prediction and the ground truth. Only the bounding boxes which have the corresponding mask value of 1 are penalized.

**Comment by bshao001** [comment]
@experiencor Thank you very much for the response. I am good at this.

**Comment by rshravan9** [comment]
Hi, 

I wrote an explanation of YOLOv2 loss function, but since MathJax is not supported here, I am sharing the link. Would be helpful to add to the wiki as well: https://stats.stackexchange.com/questions/287486/yolo-loss-function-explanation

Thanks!

**Comment by harrytruong0804** [comment]
> The mask determines what bounding boxes to be penalized. 1 for penalizing and 0 for not. For example, in YOLO, we only penalize the coordinate predictions of the bounding boxes which are responsible for predicting certain objects. So the coordinate mask has the value of 1 at the predictor bounding boxes and 0 at others. In the final loss, the mask is multiplied to the difference between the prediction and the ground truth. Only the bounding boxes which have the corresponding mask value of 1 are penalized.

@experiencor 
Could you please explain more about the mask of confidence?
The loss function contains $$l^{obj}_{i,j}$$, and $$l^{noobj}_{i,j}$$. I can not understand the $$l^{noobj}_{i,j}$$ because it is related to threshold 0.6. 
Could you elaborate on this?
